3rdparty/stout/include/stout/decompress.hpp

```diff
@@ -0,0 +1,159 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef __STOUT_DECOMPRESS_HPP__
+#define __STOUT_DECOMPRESS_HPP__
+
+#include <archive.h>
+#include <archive_entry.h>
+
+#include <glog/logging.h>
+
+#include <stout/nothing.hpp>
+#include <stout/path.hpp>
+#include <stout/try.hpp>
+
+namespace decompress {
+
+enum Format
+{
+  // These are a subset of formats supported by libarchive, but this
+  // is likely all that Mesos cares about.
+  FORMAT_ANY = 0,
+  FORMAT_BZIP2,
+  FORMAT_COMPRESS,
+  FORMAT_CPIO,
+  FORMAT_GZIP,
+  FORMAT_TAR,
+  FORMAT_ZIP
+};
+
+enum Flags
+{
+  EXTRACT_ACL = ARCHIVE_EXTRACT_ACL,
+  EXTRACT_FFLAGS = ARCHIVE_EXTRACT_FFLAGS,
+  EXTRACT_PERM = ARCHIVE_EXTRACT_PERM,
+  EXTRACT_TIME = ARCHIVE_EXTRACT_TIME,
+};
+
+inline Try<Nothing> extract(
+    const std::string& source,
+    const std::string& destination,
+    int format = Format::FORMAT_ANY,
+    int flags = Flags::EXTRACT_TIME)
+{
+  // TODO(jeffaco): Determine how to tell libarchive where to put the destination
+
+  Try<Nothing> err = Nothing();
+
+  // libdecompress, if no filename is specified, will try to read from stdin.
+  // This is one quick way to avoid that.
+  // TODO(jeffaco): Can stdin be disabled from libarchive programmatically?
+  if (source.empty())
+  {
+    return Error("No filename specified to decompress");
+  }
+
+  // Set up for generic reading of any type of compressed file
+  // TODO(jeffaco): We only support FORMAT_ANY for now. Consider if this should be
+  //     restricted to specific formats or return an error if format doesn't match.
+  //     By default, libarchive will examine the data stream to figure out what the
+  //     format is and then just use it.
+  struct archive *ar = archive_read_new();
+  archive_read_support_format_all(ar);
+  archive_read_support_filter_all(ar);
+
+  struct archive *aw = archive_write_disk_new();
+  archive_write_disk_set_options(aw, flags);
+  archive_write_disk_set_standard_lookup(aw);
+
+  // Open the compressed file for decompression
+  int r;
+  if (r = archive_read_open_filename(ar, source.c_str(), 10240))
+  {
+    err = Error(archive_error_string(ar));
+    goto error;
```

Comment:
Once the heap-allocated stuff above is RAII'ed, we can avoid the rest of the C-code too.

Reply:
* `goto error`s
* `err =` instead of `return Error()`
* `for (;;)`
* pattern match on `r` probably (e.g. `switch`)